### PR TITLE
Send a structured event to the webhooks

### PR DESCRIPTION
**Pull request by Bernardstanislas**
I am pretty sure there is a more Rust-idomatic way of doing it, I would be glad to learn how!

### CODE REVIEW

File: marketplace-infrastructure/src/event_webhook/mod.rs [hunk 1]
```diff
@@ -25,67 +25,59 @@ impl Serialize for WebhookEvent {
...
+}
+
+pub trait AggregateContext {
+	fn aggregate_name(&self) -> &'static str;
+	fn aggregate_id(&self) -> String;
+	fn event(&self) -> &'static str;
+}
+
+impl AggregateContext for WebhookEvent {
+	fn aggregate_name(&self) -> &'static str {
```
Comment (AnthonyBuisset) [inline marketplace-infrastructure/src/event_webhook/mod.rs:43]: we could factorize with the code in https://github.com/onlydustxyz/marketplace-backend/blob/main/marketplace-infrastructure/src/database/event_store/event_store.rs#L17
and
https://github.com/onlydustxyz/marketplace-backend/blob/2585d4aa878f7adf2e3dacbbf863dbf2f3058b25/marketplace-indexer/src/domain/observers/event_store.rs#L72

File: marketplace-infrastructure/src/database/event_store/event_store.rs [hunk 2]
```diff
@@ -10,23 +10,7 @@ use serde_json::Value;
 
 use crate::database::schema::events::index;
 
-trait NamedAggregate: Aggregate {
```
Comment (AnthonyBuisset) [inline marketplace-infrastructure/src/database/event_store/event_store.rs:13]: I think you need a db migration...

File: marketplace-infrastructure/src/database/event_store/event_store.rs [hunk 3]
```diff
@@ -43,7 +27,7 @@ impl<A: NamedAggregate> EventStore<A> for Client {
 				let domain_event: Event = storable_event.event.clone().into();
 				Ok(models::NewEvent {
 					timestamp: storable_event.timestamp,
-					aggregate_name: A::name(),
```
Comment (AnthonyBuisset) [inline marketplace-infrastructure/src/database/event_store/event_store.rs:46]: I think there is a mistake here, it should be `A::name()`

File: marketplace-infrastructure/src/event_webhook/mod.rs [hunk 4]
```diff
@@ -5,11 +5,43 @@ use std::env;
...
+
+impl Serialize for WebhookEvent<Event> {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: serde::Serializer,
+	{
+		let mut state = serializer.serialize_struct("Event", 4)?;
+		state.serialize_field(
+			"aggregate_type",
+			match self.0 {
```
Comment (AnthonyBuisset) [inline marketplace-infrastructure/src/event_webhook/mod.rs:33]: `Event::name()` should work as well, no `match` needed

**Review by AnthonyBuisset — CHANGES_REQUESTED** [review]

File: marketplace-infrastructure/src/event_webhook/mod.rs (hunk 1 above)
Comment (Bernardstanislas) [reply marketplace-infrastructure/src/event_webhook/mod.rs:43]: As seen IRL, it is not a good idea to mix these two concepts

File: marketplace-infrastructure/src/event_webhook/mod.rs [hunk 5]
```diff
@@ -5,11 +5,82 @@ use std::env;
...
+impl Serialize for WebhookEvent {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: serde::Serializer,
+	{
+		let mut state = serializer.serialize_struct("Event", 4)?;
+		state.serialize_field("aggregate_type", self.aggregate_name())?;
+		state.serialize_field("aggregate_id", &self.aggregate_id())?;
+		state.serialize_field("event", &self.event())?;
+		state.serialize_field("payload", &self.0)?;
```
[the remark this answers is not in the document]
Comment (Bernardstanislas) [reply marketplace-infrastructure/src/event_webhook/mod.rs]: Agreed, refactored

File: marketplace-infrastructure/src/event_webhook/mod.rs [hunk 6]
```diff
@@ -6,10 +6,53 @@ use std::env;
...
+		Self(event)
+	}
+}
+
+impl Serialize for WebhookEvent {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: Serializer,
+	{
+		let mut state = serializer.serialize_struct("Event", 4)?;
```
Comment (AnthonyBuisset) [inline marketplace-infrastructure/src/event_webhook/mod.rs:28]: still 4 ?

**Review by AnthonyBuisset — APPROVED** [review]

File: marketplace-infrastructure/src/event_webhook/mod.rs (hunk 6 above)
Comment (Bernardstanislas) [reply marketplace-infrastructure/src/event_webhook/mod.rs:28]: nooo 3 thank you

**Review by tdelabro — APPROVED** [review]

**Pull request merged**